src/BotPlutusInterface/Balance.hs

```diff
@@ -156,7 +216,38 @@ balanceTxIO pabConf ownPkh unbalancedTx =
 
       if balancedTx == tx
         then pure (balancedTx, minUtxos)
-        else loop utxoIndex privKeys minUtxos balancedTx
+        else balanceTxLoop utxoIndex privKeys minUtxos balancedTx
+
+-- `utxosAndCollateralAtAddress` returns all the utxos that can be used as input of a `Tx`,
+-- i.e. we filter out `CollateralUtxo` present at the user's address, so it can't be used as input of a `Tx`.
+-- This function throws error if the `Tx` type is of `BalanceTxWithScripts` but there's not `CollateralUtxo`
+-- in the environment.
+utxosAndCollateralAtAddress ::
+  forall (w :: Type) (effs :: [Type -> Type]).
+  Member (PABEffect w) effs =>
+  BalanceTx ->
+  PABConfig ->
+  Address ->
+  Eff effs (Either Text (Map TxOutRef Tx.ChainIndexTxOut, Maybe CollateralUtxo))
+utxosAndCollateralAtAddress balanceTxType pabConf changeAddr =
+  runEitherT $ do
+    utxos <- newEitherT $ CardanoCLI.utxosAt @w pabConf changeAddr
+    inMemCollateral <- lift $ getInMemCollateral @w
+
+    case balanceTxType of
+      BalanceTxCollateral -> return (utxos, Nothing)
+      BalanceTxWithoutScripts -> return (removeCollateralFromMap inMemCollateral utxos, Nothing)
+      BalanceTxWithScripts ->
+        case inMemCollateral of
+          Nothing -> throwE "The given transaction uses script, but there's no collateral provided. This usually means that, we failed to create Tx and update our ContractEnvironment."
+          (Just _) -> return (removeCollateralFromMap inMemCollateral utxos, inMemCollateral)
```

Comment:
Redundant parens here (and a `return` instead of `pure`)

Reply:
Fixed